contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,245 @@
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "mysql_pool")]
+pub extern crate mysql;
+#[cfg(feature = "mysql_pool")]
+pub extern crate r2d2_mysql;
+
+#[cfg(feature = "sqlite_pool")]
+pub extern crate rusqlite;
+#[cfg(feature = "sqlite_pool")]
+pub extern crate r2d2_sqlite;
+
+#[cfg(feature = "cypher_pool")]
+pub extern crate rusted_cypher;
+#[cfg(feature = "cypher_pool")]
+pub extern crate r2d2_cypher;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+///
+/// Example configuration:
+/// ```toml
+/// [[global.databases]]
+/// name = "my_database"
+/// url = "postgres://postgres@localhost/my_database
+/// pool_size = 4
+/// ```
+pub struct DatabaseConfig<'a> {
+    /// The connection URL specified in the Rocket configuration
+    pub url: &'a str,
+    /// The size of the pool to be initialized. Defaults to the number of Rocket workers.
+    pub pool_size: u32,
+    /// Any extra options that are included in the configuration.
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// The error type for fetching the DatabaseConfig
+#[derive(Debug)]
+pub enum DatabaseConfigError {
+    MissingTable,
+    MalformedConfiguration,
+    MissingConnectionUrl,
+    InvalidConnectionUrl,
+    InvalidPoolSize(i64),
+}
+
+impl<'a> Display for DatabaseConfigError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            DatabaseConfigError::MissingTable => {
+                write!(f, "Database-specific configuration was not found for this database")
+            },
+            DatabaseConfigError::MalformedConfiguration => {
+                write!(f, "The configuration for this database is malformed")
+            }
+            DatabaseConfigError::MissingConnectionUrl => {
+                write!(f, "The connection URL is missing for this database")
+            },
+            DatabaseConfigError::InvalidConnectionUrl => {
+                write!(f, "The specified connection URL is invalid")
+            },
+            DatabaseConfigError::InvalidPoolSize(invalid_size) => {
+                write!(f, "'{}' is not a valid value for `pool_size`", invalid_size)
+            },
+        }
+    }
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
```

Comment:
Let's get a real, tested example: no ignoring. Use `rocket::ignite()` then an attach fairing (all of this hidden) to get an instance of `rocket` on which you can actually call `config()` to call this method. To actually test the config, create a custom config, pass it into `rocket::custom()`, and proceed as normal. Something like:

## Example

Given an `&Config` derived from the following configuration:

```toml
[global.databases]
my_db = { url = "db/db.sqlite", pool_size = 25 }
yet_another_db = { url = "db/other.sqlite" }
```

```rust
use rocket_contrib::databases::database_config;

# let my_db = bmap! { "url" => "db/db.sqlite" };
# let yet_another_db = bmap! { "url" => "db/other.sqlite" };
# let databases = bmap! { "my_db" => my_db, "yet_another_db" => yet_another_db);
# let config = Config::build(Environment::Debug).extra("databases", databases);
# rocket::custom(config).attach(AdHoc::on_attach(|rocket| {
let config = database_config("my_db", rocket.config()).expect("my_db config okay");
assert_eq!(config.url, "db/db.sqlite");
assert_eq!(config.pool_size, 25);

let config = database_config("yet_another_db", rocket.config()).expect("yet_another_db config okay");
assert_eq!(config.url, "db/other.sqlite");

let error = database_config("unknown", rocket.config()).unwrap_err();
assert_eq!(error, DatabaseConfigError::MissingKey);
# });
```

Note how I've ignored all of the setup code. The nice this is that because this is an attach fairing, this will run and test everything, for real.